{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Computational Workflows for biomedical data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Welcome to the course Computational Workflows for Biomedical Data. Over the next two weeks, you will learn how to leverage nf-core pipelines to analyze biomedical data and gain hands-on experience in creating your own pipelines, with a strong emphasis on Nextflow and nf-core.\n",
    "\n",
    "Course Structure:\n",
    "\n",
    "- Week 1: You will use a variety of nf-core pipelines to analyze a publicly available biomedical study.\n",
    "- Week 2: We will shift focus to learning the basics of Nextflow, enabling you to design and implement your own computational workflows.<br>\n",
    "- Final Project: The last couple of days, you will apply your knowledge to create a custom pipeline for analyzing biomedical data using Nextflow and the nf-core template."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basics\n",
    "\n",
    "If you have not installed all required software, please do so now asap!\n",
    "\n",
    "\n",
    "If you already installed all software, please go on and start answering the questions in this notebook. If you have any questions, don't hesitate to approach us."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. What is nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. How many pipelines are there currently in nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Are there any non-bioinformatic pipelines in nf-core?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Let's go back a couple of steps. What is a pipeline and what do we use it for?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Why do you think nf-core adheres to strict guidelines?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "6. What are the main features of nf-core pipelines?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "source": [
    "## Let's start using the pipelines"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Find the nf-core pipeline used to measure differential abundance of genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025/notebooks/day_01\n"
     ]
    }
   ],
   "source": [
    "# run the pipeline in a cell \n",
    "# to run bash in jupyter notebooks, simply use ! before the command\n",
    "# e.g.\n",
    "\n",
    "!pwd\n",
    "\n",
    "\n",
    "# For the tasks in the first week, please use the command line to run your commands and simply paste the commands you used in the respective cells!\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# run the pipeline in the test profile using docker containers\n",
    "# make sure to specify the version you want to use (use the latest one)\n",
    "\n",
    "\n",
    "!nextflow run nf-core/differentialabundance      -profile test,docker      --outdir differentialout  \n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\"\"\"\n",
    " N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [sharp_bhabha] DSL2 - revision: 3dd360fed0 [master]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : master\n",
    "  runName                     : sharp_bhabha\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025\n",
    "  workDir                     : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025/work\n",
    "  projectDir                  : /Users/peterbrederlow/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : peterbrederlow\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : differentialout\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "[-        ] NFCâ€¦ENTIALABUNDANCE:GUNZIP_GTF -\n",
    "executor >  local (1)\n",
    "executor >  local (1)\n",
    "executor >  local (2)\n",
    "executor >  local (2)\n",
    "executor >  local (2)\n",
    "executor >  local (3)\n",
    "executor >  local (21)\n",
    "[3d/6c8107] NFCâ€¦musculus.GRCm38.81.gtf.gz) | 1 of 1 âœ”\n",
    "[fb/024b3a] NFCâ€¦TF_TO_TABLE (Mus_musculus) | 1 of 1 âœ”\n",
    "[ec/a18c38] NFCâ€¦SRP254919.samplesheet.csv) | 1 of 1 âœ”\n",
    "[e2/c6322c] NFCâ€¦RIXFILTER ([id:SRP254919]) | 1 of 1 âœ”\n",
    "[4a/54a610] NFCâ€¦, target:hND6, blocking:]) | 1 of 1 âœ”\n",
    "[e5/7a5bf2] NFCâ€¦, target:hND6, blocking:]) | 2 of 2 âœ”\n",
    "[fa/5e1a92] NFCâ€¦DANCE:FILTER_DIFFTABLE (2) | 2 of 2 âœ”\n",
    "[1e/d9102c] NFCâ€¦ (treatment_mCherry_hND6_) | 1 of 1 âœ”\n",
    "[4e/8fa6af] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[ef/5b59cc] NFCâ€¦E:TABULAR_TO_GSEA_CHIP (1) | 1 of 1 âœ”\n",
    "[ae/4248f7] NFCâ€¦ (treatment_mCherry_hND6_) | 2 of 2 âœ”\n",
    "[86/719a57] NFCâ€¦OT_EXPLORATORY (treatment) | 1 of 1 âœ”\n",
    "[32/4c1e72] NFCâ€¦ (treatment_mCherry_hND6_) | 2 of 2 âœ”\n",
    "[11/cfc74f] NFCâ€¦E:SHINYNGS_APP (SRP254919) | 1 of 1 âœ”\n",
    "[b4/2a836f] NFCâ€¦RKDOWNNOTEBOOK (SRP254919) | 1 of 1 âœ”\n",
    "[6d/19ee6f] NFCâ€¦_REPORT_BUNDLE (SRP254919) | 1 of 1 âœ”\n",
    "Completed at: 29-Sep-2025 12:18:27\n",
    "Duration    : 1m 40s\n",
    "CPU hours   : (a few seconds)\n",
    "Succeeded   : 21\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# repeat the run. What did change?\n",
    "!nextflow run nf-core/differentialabundance      -profile test,docker      --outdir differentialout  \n",
    "\"\"\"\n",
    " N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [nauseous_planck] DSL2 - revision: 3dd360fed0 [master]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : master\n",
    "  runName                     : nauseous_planck\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025\n",
    "  workDir                     : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025/work\n",
    "  projectDir                  : /Users/peterbrederlow/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : peterbrederlow\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : differentialout\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "[-        ] NFCâ€¦ENTIALABUNDANCE:GUNZIP_GTF -\n",
    "executor >  local (1)\n",
    "executor >  local (1)\n",
    "executor >  local (2)\n",
    "executor >  local (2)\n",
    "executor >  local (3)\n",
    "executor >  local (3)\n",
    "executor >  local (6)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (10)\n",
    "executor >  local (21)\n",
    "[1e/eaf2ad] NFCâ€¦musculus.GRCm38.81.gtf.gz) | 1 of 1 âœ”\n",
    "[e7/f44672] NFCâ€¦TF_TO_TABLE (Mus_musculus) | 1 of 1 âœ”\n",
    "[0c/acaba9] NFCâ€¦SRP254919.samplesheet.csv) | 1 of 1 âœ”\n",
    "[79/d5f4df] NFCâ€¦RIXFILTER ([id:SRP254919]) | 1 of 1 âœ”\n",
    "[c0/7cf425] NFCâ€¦, target:hND6, blocking:]) | 1 of 1 âœ”\n",
    "[08/bc0f4f] NFCâ€¦, target:hND6, blocking:]) | 2 of 2 âœ”\n",
    "[80/643a36] NFCâ€¦DANCE:FILTER_DIFFTABLE (2) | 2 of 2 âœ”\n",
    "[7c/5911d4] NFCâ€¦ (treatment_mCherry_hND6_) | 1 of 1 âœ”\n",
    "[a5/1f858b] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[c1/b7f565] NFCâ€¦E:TABULAR_TO_GSEA_CHIP (1) | 1 of 1 âœ”\n",
    "[39/269046] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[cd/02d30c] NFCâ€¦OT_EXPLORATORY (treatment) | 1 of 1 âœ”\n",
    "[e8/dec911] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[f6/d38cb8] NFCâ€¦E:SHINYNGS_APP (SRP254919) | 1 of 1 âœ”\n",
    "[a6/1da6d7] NFCâ€¦RKDOWNNOTEBOOK (SRP254919) | 1 of 1 âœ”\n",
    "[ac/1ec297] NFCâ€¦_REPORT_BUNDLE (SRP254919) | 1 of 1 âœ”\n",
    "Completed at: 29-Sep-2025 12:38:08\n",
    "Duration    : 1m 30s\n",
    "CPU hours   : (a few seconds)\n",
    "Succeeded   : 21\n",
    "\"\"\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# now set -resume to the command. What did change?\n",
    "!nextflow run nf-core/differentialabundance      -profile test,docker      --outdir differentialout  -resume\n",
    "\n",
    "\"\"\" \n",
    "N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [astonishing_darwin] DSL2 - revision: 3dd360fed0 [master]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : master\n",
    "  runName                     : astonishing_darwin\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025\n",
    "  workDir                     : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025/work\n",
    "  projectDir                  : /Users/peterbrederlow/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : peterbrederlow\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : differentialout\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "executor >  local (1)\n",
    "[01/e88459] NFCâ€¦musculus.GRCm38.81.gtf.gz) | 1 of 1, cached: 1 âœ”\n",
    "[60/b20b88] NFCâ€¦TF_TO_TABLE (Mus_musculus) | 1 of 1, cached: 1 âœ”\n",
    "[6b/73e96b] NFCâ€¦SRP254919.samplesheet.csv) | 1 of 1, cached: 1 âœ”\n",
    "[63/0f726b] NFCâ€¦RIXFILTER ([id:SRP254919]) | 1 of 1, cached: 1 âœ”\n",
    "[5b/11ddc3] NFCâ€¦, target:hND6, blocking:]) | 1 of 1, cached: 1 âœ”\n",
    "[77/e19a58] NFCâ€¦, target:hND6, blocking:]) | 2 of 2, cached: 2 âœ”\n",
    "[b0/1e8af6] NFCâ€¦DANCE:FILTER_DIFFTABLE (1) | 2 of 2, cached: 2 âœ”\n",
    "[69/6dfd18] NFCâ€¦ (treatment_mCherry_hND6_) | 1 of 1, cached: 1 âœ”\n",
    "[5b/9669b3] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2, cached: 2 âœ”\n",
    "[f6/d358c8] NFCâ€¦E:TABULAR_TO_GSEA_CHIP (1) | 1 of 1, cached: 1 âœ”\n",
    "[bb/d70cec] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2, cached: 2 âœ”\n",
    "[ce/8e5462] NFCâ€¦OT_EXPLORATORY (treatment) | 1 of 1, cached: 1 âœ”\n",
    "[b8/3b27e3] NFCâ€¦ (treatment_mCherry_hND6_) | 2 of 2, cached: 2 âœ”\n",
    "[52/b7b097] NFCâ€¦E:SHINYNGS_APP (SRP254919) | 1 of 1 âœ”\n",
    "[d8/89bbd2] NFCâ€¦RKDOWNNOTEBOOK (SRP254919) | 1 of 1, cached: 1 âœ”\n",
    "[23/0d0bdd] NFCâ€¦_REPORT_BUNDLE (SRP254919) | 1 of 1, cached: 1 âœ”\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Check out the current directory. Next to the outdir you specified, what else has changed?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# delete the work directory and run the pipeline again using -resume. What did change?\n",
    "\"\"\"(nf-env) peterbrederlow@MacBook-Pro-von-Peter computational-workflows-2025 % nextflow run nf-core/differentialabundance      -profile test,docker      --outdir differentialout -resume\n",
    "\n",
    " N E X T F L O W   ~  version 25.04.7\n",
    "\n",
    "Launching `https://github.com/nf-core/differentialabundance` [nauseous_planck] DSL2 - revision: 3dd360fed0 [master]\n",
    "\n",
    "WARN: Access to undefined parameter `monochromeLogs` -- Initialise it to a default value eg. `params.monochromeLogs = some_value`\n",
    "\n",
    "\n",
    "------------------------------------------------------\n",
    "                                        ,--./,-.\n",
    "        ___     __   __   __   ___     /,-._.--~'\n",
    "  |\\ | |__  __ /  ` /  \\ |__) |__         }  {\n",
    "  | \\| |       \\__, \\__/ |  \\ |___     \\`-._,-`-,\n",
    "                                        `._,._,'\n",
    "  nf-core/differentialabundance v1.5.0-g3dd360f\n",
    "------------------------------------------------------\n",
    "Core Nextflow options\n",
    "  revision                    : master\n",
    "  runName                     : nauseous_planck\n",
    "  containerEngine             : docker\n",
    "  container                   : [RMARKDOWNNOTEBOOK:biocontainers/r-shinyngs:1.8.8--r43hdfd78af_0]\n",
    "  launchDir                   : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025\n",
    "  workDir                     : /Users/peterbrederlow/Documents/Uni/MasterBioinformatik/SS25/WorkflowCourse/computational-workflows-2025/work\n",
    "  projectDir                  : /Users/peterbrederlow/.nextflow/assets/nf-core/differentialabundance\n",
    "  userName                    : peterbrederlow\n",
    "  profile                     : test,docker\n",
    "  configFiles                 : \n",
    "\n",
    "Input/output options\n",
    "  study_name                  : SRP254919\n",
    "  input                       : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv\n",
    "  contrasts                   : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.contrasts.csv\n",
    "  outdir                      : differentialout\n",
    "\n",
    "Abundance values\n",
    "  matrix                      : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.salmon.merged.gene_counts.top1000cov.tsv\n",
    "  transcript_length_matrix    : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.spoofed_lengths.tsv\n",
    "  affy_cel_files_archive      : null\n",
    "  querygse                    : null\n",
    "\n",
    "Observations (e.g. samples) options\n",
    "  observations_name_col       : sample\n",
    "\n",
    "Affy input options\n",
    "  affy_cdfname                : null\n",
    "\n",
    "Filtering\n",
    "  filtering_min_abundance     : 10\n",
    "\n",
    "Exploratory analysis\n",
    "  exploratory_main_variable   : contrasts\n",
    "  exploratory_log2_assays     : raw,normalised\n",
    "\n",
    "Differential analysis\n",
    "  differential_file_suffix    : .deseq2.results.tsv\n",
    "\n",
    "DESeq2 specific options (RNA-seq only)\n",
    "  deseq2_vst_nsub             : 500\n",
    "\n",
    "Limma specific options (microarray only)\n",
    "  limma_spacing               : null\n",
    "  limma_block                 : null\n",
    "  limma_correlation           : null\n",
    "\n",
    "GSEA\n",
    "  gsea_run                    : true\n",
    "\n",
    "gprofiler2\n",
    "  gprofiler2_correction_method: gSCS\n",
    "  gprofiler2_background_file  : auto\n",
    "\n",
    "Shiny app settings\n",
    "  shinyngs_shinyapps_account  : null\n",
    "  shinyngs_shinyapps_app_name : null\n",
    "\n",
    "Options related to gene set analysis\n",
    "  gene_sets_files             : https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/gene_set_analysis/mh.all.v2022.1.Mm.symbols.gmt\n",
    "\n",
    "Reporting options\n",
    "  report_title                : null\n",
    "  report_author               : null\n",
    "  report_contributors         : Jane Doe\n",
    "Director of Institute of Microbiology\n",
    "University of Smallville;John Smith\n",
    "PhD student\n",
    "Institute of Microbiology\n",
    "University of Smallville\n",
    "  report_description          : null\n",
    "  report_round_digits         : 3\n",
    "\n",
    "Reference genome options\n",
    "  gtf                         : https://ftp.ensembl.org/pub/release-81/gtf/mus_musculus/Mus_musculus.GRCm38.81.gtf.gz\n",
    "\n",
    "Institutional config options\n",
    "  config_profile_name         : Test profile\n",
    "  config_profile_description  : Minimal test dataset to check pipeline function\n",
    "\n",
    "Max job request options\n",
    "  max_cpus                    : 2\n",
    "  max_memory                  : 6.GB\n",
    "  max_time                    : 6.h\n",
    "\n",
    "!! Only displaying parameters that differ from the pipeline defaults !!\n",
    "------------------------------------------------------\n",
    "If you use nf-core/differentialabundance for your analysis please cite:\n",
    "\n",
    "* The pipeline\n",
    "  https://doi.org/10.5281/zenodo.7568000\n",
    "\n",
    "* The nf-core framework\n",
    "  https://doi.org/10.1038/s41587-020-0439-x\n",
    "\n",
    "* Software dependencies\n",
    "  https://github.com/nf-core/differentialabundance/blob/master/CITATIONS.md\n",
    "------------------------------------------------------\n",
    "[-        ] NFCâ€¦ENTIALABUNDANCE:GUNZIP_GTF -\n",
    "executor >  local (1)\n",
    "executor >  local (1)\n",
    "executor >  local (2)\n",
    "executor >  local (2)\n",
    "executor >  local (3)\n",
    "executor >  local (3)\n",
    "executor >  local (6)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (7)\n",
    "executor >  local (10)\n",
    "executor >  local (21)\n",
    "[1e/eaf2ad] NFCâ€¦musculus.GRCm38.81.gtf.gz) | 1 of 1 âœ”\n",
    "[e7/f44672] NFCâ€¦TF_TO_TABLE (Mus_musculus) | 1 of 1 âœ”\n",
    "[0c/acaba9] NFCâ€¦SRP254919.samplesheet.csv) | 1 of 1 âœ”\n",
    "[79/d5f4df] NFCâ€¦RIXFILTER ([id:SRP254919]) | 1 of 1 âœ”\n",
    "[c0/7cf425] NFCâ€¦, target:hND6, blocking:]) | 1 of 1 âœ”\n",
    "[08/bc0f4f] NFCâ€¦, target:hND6, blocking:]) | 2 of 2 âœ”\n",
    "[80/643a36] NFCâ€¦DANCE:FILTER_DIFFTABLE (2) | 2 of 2 âœ”\n",
    "[7c/5911d4] NFCâ€¦ (treatment_mCherry_hND6_) | 1 of 1 âœ”\n",
    "[a5/1f858b] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[c1/b7f565] NFCâ€¦E:TABULAR_TO_GSEA_CHIP (1) | 1 of 1 âœ”\n",
    "[39/269046] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[cd/02d30c] NFCâ€¦OT_EXPLORATORY (treatment) | 1 of 1 âœ”\n",
    "[e8/dec911] NFCâ€¦Cherry_hND6_sample_number) | 2 of 2 âœ”\n",
    "[f6/d38cb8] NFCâ€¦E:SHINYNGS_APP (SRP254919) | 1 of 1 âœ”\n",
    "[a6/1da6d7] NFCâ€¦RKDOWNNOTEBOOK (SRP254919) | 1 of 1 âœ”\n",
    "[ac/1ec297] NFCâ€¦_REPORT_BUNDLE (SRP254919) | 1 of 1 âœ”\n",
    "Completed at: 29-Sep-2025 12:38:08\n",
    "Duration    : 1m 30s\n",
    "CPU hours   : (a few seconds)\n",
    "Succeeded   : 21\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What changed?\n",
    "Now the pipeline ran similar to the pipeline without the resume flag"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lets look at the results"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### What is differential abundance analysis?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Give the most important plots from the report:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
